python/pyspark/sql/types.py

```diff
@@ -1160,18 +1150,22 @@ def _has_struct_or_date(dt):
 def _create_properties(fields):
     """Create properties according to fields"""
```

Comment:
basically my worry is that there is a lot of magic happening with the python implementation, and it is very hard to figure out what's going on without extensive comments.